### ISSUE TITLE
Possible failure to release resources on Android with react-native SDK, Strict Mode crashes, warnings about writing from main thread

### ISSUE THREAD

**Issue by AndrewMorsillo** [smartlook/smartlook-mobile-issue-tracker#20]
**Platform**
Android, iOS

**Technology**
React Native

**Rendering [optional]**
N/A

**Affected devices [optional]**
N/A

**Smartlook SDK version**
"smartlook-react-native-wrapper": "^0.34.0"

**Summary**
I started my React Native Android app in[ Android strict mode ](https://developer.android.com/reference/android/os/StrictMode) to check on an issue. It reported what is apparently Smartlook failing to properly release resources.

Side note: I also had to disable `detectDiskReads()` and `detectDiskWrites()` from StrictMode since they constantly spammed the console due to Smartlook. It appears the Smartlook SDK does a lot of disk read/write on main thread against Android recommendations.

**Steps to reproduce [optional]**
1. Add smartlook to react-native app
2. Turn on Android strict mode (add to MainApplication.java) 
```
    StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
//            .detectDiskReads()
//            .detectDiskWrites()
            .detectNetwork()   // or .detectAll() for all detectable problems
            .penaltyLog()
            .build());
    StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
            .detectLeakedSqlLiteObjects()
            .detectLeakedClosableObjects()
            .penaltyLog()
            .penaltyDeath()
            .build());
```
3. `react-native run-android --variant=release` and observe logged output with adb

**Relevant logs and/or screenshots**
```
2020-10-15 13:08:59.805 14555-14571/com.whetware.taskhero D/StrictMode: StrictMode policy violation: android.os.strictmode.LeakedClosableViolation: A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
        at android.os.StrictMode$AndroidCloseGuardReporter.report(StrictMode.java:1924)
        at dalvik.system.CloseGuard.warnIfOpen(CloseGuard.java:303)
        at android.hardware.SystemSensorManager$BaseEventQueue.dispose(SystemSensorManager.java:733)
        at android.hardware.SystemSensorManager$BaseEventQueue.finalize(SystemSensorManager.java:724)
        at java.lang.Daemons$FinalizerDaemon.doFinalize(Daemons.java:291)
        at java.lang.Daemons$FinalizerDaemon.runInternal(Daemons.java:278)
        at java.lang.Daemons$Daemon.run(Daemons.java:139)
        at java.lang.Thread.run(Thread.java:923)
     Caused by: java.lang.Throwable: Explicit termination method 'dispose' not called
        at dalvik.system.CloseGuard.openWithCallSite(CloseGuard.java:259)
        at dalvik.system.CloseGuard.open(CloseGuard.java:230)
        at android.hardware.SystemSensorManager$BaseEventQueue.<init>(SystemSensorManager.java:655)
        at android.hardware.SystemSensorManager$SensorEventQueue.<init>(SystemSensorManager.java:784)
        at android.hardware.SystemSensorManager.registerListenerImpl(SystemSensorManager.java:178)
        at android.hardware.SensorManager.registerListener(SensorManager.java:823)
        at android.hardware.SensorManager.registerListener(SensorManager.java:730)
        at android.view.OrientationEventListener.enable(OrientationEventListener.java:93)
        at a.b.a.a.e.i.b.b.h(SourceFile:11)
        at a.b.a.a.e.i.b.b.f(SourceFile:14)
        at a.b.a.a.e.i.b.b.a(SourceFile:3)
        at a.b.a.a.e.i.b.b$n.run(SourceFile:1)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
2020-10-15 13:08:59.805 14555-14571/com.whetware.taskhero W/System.err: StrictMode VmPolicy violation with POLICY_DEATH; shutting down.
```

**Comment by adamblack** [comment]
Hi @AndrewMorsillo ,

thanks for info. We will look into that (it might be connected to issue that you had reported some time ago, but we are still having issues of how to emulate that on our side :/ ).

We do not read/write on UI thread but we will try that for sure, however it is true that we cache some info to internal app's storage. -> In probably next version, this is gonna be refactored/removed. (We are already working on it)

**Comment by AndrewMorsillo** [comment]
@adamblack Thanks for the update. Hopefully that can clear up the log spam from https://github.com/smartlook/smartlook-mobile-issue-tracker/issues/3 as well 

Perhaps this is related and I'm not sure if I shared this log with you yet but there are a lot of `E/MPEG4Writer: Stop() called but track is not started or stopped` being logged. This seems to get logged immediately before a few thousand lines of "invalid id" show up.

When the app is at rest doing nothing the logs generally look like the following repeated every couple seconds. I hope this can give you more of a clue about the issue.

```
020-10-16 20:16:45.152 30384-31276/com.whetware.taskhero D/MPEG4Writer: PreAllocation enabled
2020-10-16 20:16:45.152 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x00000061.
2020-10-16 20:16:45.152 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x00000063.
2020-10-16 20:16:45.152 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x00001ebd.
2020-10-16 20:16:45.152 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x00001dbd.
2020-10-16 20:16:45.152 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x00001eb9.
2020-10-16 20:16:45.152 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x00001eb7.
2020-10-16 20:16:45.152 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x00001eb3.

....a few thousand "invalid id"

2020-10-16 20:16:45.161 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x000002dd.
2020-10-16 20:16:45.161 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x000002eb.
2020-10-16 20:16:45.161 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x000002e3.
2020-10-16 20:16:45.161 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x000002e9.
2020-10-16 20:16:45.164 30384-3927/com.whetware.taskhero I/OMXClient: IOmx service obtained
2020-10-16 20:16:45.170 30384-3927/com.whetware.taskhero W/OMXUtils: do not know color format 0x7f000789 = 2130708361
2020-10-16 20:16:45.172 30384-3927/com.whetware.taskhero I/ACodec: setupAVCEncoderParameters with [profile: Baseline] [level: Level41]
2020-10-16 20:16:45.173 30384-3927/com.whetware.taskhero I/ACodec: [OMX.google.h264.encoder] cannot encode color aspects. Ignoring.
2020-10-16 20:16:45.174 30384-3927/com.whetware.taskhero I/ACodec: [OMX.google.h264.encoder] cannot encode HDR static metadata. Ignoring.
2020-10-16 20:16:45.174 30384-3927/com.whetware.taskhero I/ACodec: setupVideoEncoder succeeded
2020-10-16 20:16:45.174 30384-3927/com.whetware.taskhero W/OMXUtils: do not know color format 0x7f000789 = 2130708361
2020-10-16 20:16:45.224 30384-3927/com.whetware.taskhero D/ACodec: dataspace changed to 0x10c40000 (R:2(Limited), P:4(BT601_6_525), M:3(BT601_6), T:3(SMPTE170M)) (R:2(Limited), S:4(BT601_525), T:3(SMPTE_170M))
2020-10-16 20:16:45.226 30384-31276/com.whetware.taskhero D/MPEG4Writer: fpathconf _PC_FILESIZEBITS:64
2020-10-16 20:16:45.226 30384-31276/com.whetware.taskhero D/MPEG4Writer: File size limit set to 4503599627370495 bytes implicitly
2020-10-16 20:16:45.226 30384-31276/com.whetware.taskhero D/MPEG4Writer: MP4WtrCtrlHlpLooper Started
2020-10-16 20:16:45.227 30384-31276/com.whetware.taskhero I/MPEG4Writer: limits: 4503599627370495/0 bytes/us, bit rate: -1 bps and the estimated moov size 3192 bytes
2020-10-16 20:16:45.227 30384-3933/com.whetware.taskhero D/MPEG4Writer: kWhatNoIOErrorSoFar
2020-10-16 20:16:45.299 30384-3934/com.whetware.taskhero I/MPEG4Writer: setStartTimestampUs: 0
2020-10-16 20:16:45.299 30384-3934/com.whetware.taskhero I/MPEG4Writer: Earliest track starting time: 0
2020-10-16 20:16:45.650 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x00000015.
2020-10-16 20:16:45.650 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x00000025.
2020-10-16 20:16:45.650 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x00000057.
2020-10-16 20:16:45.650 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x00000055.
2020-10-16 20:16:45.650 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x00000053.
2020-10-16 20:16:45.650 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x0000004f.
2020-10-16 20:16:45.650 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x00000031.

....a few thousand more "invalid id"

2020-10-16 20:16:46.160 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x000002df.
2020-10-16 20:16:46.160 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x000002dd.
2020-10-16 20:16:46.160 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x000002eb.
2020-10-16 20:16:46.160 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x000002e3.
2020-10-16 20:16:46.160 30384-3925/com.whetware.taskhero E/etware.taskher: Invalid ID 0x000002e9.
2020-10-16 20:16:46.441 30384-31276/com.whetware.taskhero D/MPEG4Writer: reset()
2020-10-16 20:16:46.441 30384-31276/com.whetware.taskhero D/MPEG4Writer: Video track stopping. Stop source
2020-10-16 20:16:46.441 30384-31276/com.whetware.taskhero D/MPEG4Writer: Video track source stopping
2020-10-16 20:16:46.441 30384-31276/com.whetware.taskhero D/MPEG4Writer: Video track source stopped
2020-10-16 20:16:46.441 30384-3934/com.whetware.taskhero V/MediaWriter: Track event err/info msg:101, trackId:1, type:1000,val:-1011
2020-10-16 20:16:46.441 30384-3934/com.whetware.taskhero I/MPEG4Writer: Received total/0-length (33/0) buffers and encoded 33 frames. - Video
2020-10-16 20:16:46.441 30384-31276/com.whetware.taskhero D/MPEG4Writer: Video track stopped. Status:0. Stop source
2020-10-16 20:16:46.441 30384-3932/com.whetware.taskhero D/MPEG4Writer: 0 chunks are written in the last batch
2020-10-16 20:16:46.442 30384-31276/com.whetware.taskhero D/MPEG4Writer: WriterThread stopped. Status:0
2020-10-16 20:16:46.443 30384-31276/com.whetware.taskhero I/MPEG4Writer: Adjust the moov start time from 0 us -> 0 us
2020-10-16 20:16:46.443 30384-31276/com.whetware.taskhero I/MPEG4Writer: MOOV atom was written to the file
2020-10-16 20:16:46.443 30384-31276/com.whetware.taskhero D/MPEG4Writer: release()
2020-10-16 20:16:46.443 30384-31276/com.whetware.taskhero D/MPEG4Writer: ftruncate mPreAllocateFileEndOffset:161872 mOffset:1182 mMdatEndOffset:160072 diff:1800
2020-10-16 20:16:46.446 30384-31276/com.whetware.taskhero D/MPEG4Writer: MP4WtrCtrlHlpLooper stopped
2020-10-16 20:16:46.446 30384-31276/com.whetware.taskhero D/MPEG4Writer: Top 5 write durations(microseconds): #1:46 #2:55 #3:169 #4:716 #5:2750
2020-10-16 20:16:46.446 30384-31276/com.whetware.taskhero D/MPEG4Writer: reset()
2020-10-16 20:16:46.446 30384-31276/com.whetware.taskhero D/MPEG4Writer: Video track stopping. Stop source
2020-10-16 20:16:46.446 30384-31276/com.whetware.taskhero E/MPEG4Writer: Stop() called but track is not started or stopped

.... thousands more "invalid id" then these logs repeat
```

**Comment by adamblack** [comment]
Hi @AndrewMorsillo ,

we found the cause of the issue. We are going to release new version soon - in worst case, next week.

We are using some Android callbacks and for some reason, RN does not like that. We will make sure it is fixed however I believe this should be later fixed on RN side as well.

**Comment by AndrewMorsillo** [comment]
Glad to hear it, thanks!

**Comment by adamblack** [comment]
Hi @AndrewMorsillo ,

we just released new version 0.40.0 - for testing I used project you provided to us (written in typescript). Seems to be ok, would be great to get your feedback before we somehow close this issue. :]
